x-pack/plugins/endpoint/server/routes/resolver/utils/tree.ts

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { ResolverEvent, Node, NodeStats, NodePagination } from '../../../../common/types';
+import { extractEntityID, extractParentEntityID } from './normalize';
+import { buildPaginationCursor } from './pagination';
+
+type ExtractFunction = (event: ResolverEvent) => string | undefined;
+
+function createNode(id: string): Node {
+  return { id, children: [], pagination: {}, events: [], alerts: [], lifecycle: [] };
+}
+
+// This class aids in constructing a tree of process events. It works in the following way:
+//
+// 1. We construct a tree structure starting with the root node for the event we're requesting.
+// 2. We leverage the ability to pass hashes and arrays by reference to construct a fast cache of
+//    process identifiers that updates the tree structure as we push values into the cache.
+//
+// When we query a single level of results for child process events we have a flattened, sorted result
+// list that we need to add into a constructed tree. We also need to signal in an API response whether
+// or not there are more child processes events that we have not yet retrieved, and, if so, for what parent
```

Comment:
Why do we need to signal in an API response whether or not there are more child process events? Even if we needed that, how would we calculate it?

There could be more events if:
* If there are still events on the endpoint itself that haven't been sent
* There are events in an intermediate queueing system e.g. Logstash
* There are events that have been received by ES but not yet propagated
* The endpoint hasn't been turned off, rebooted, or had the sensor removed, and at least 1 process in the graph hasn't been terminated.

Unless this API can definitely say that there are no more events, the UI will need to let the user check for events.

Reply:
It might still be useful to have the API tell you that either there definitively are more events or that there were no more last time it looked. Maybe to the user the difference the difference would be a "get more events" button vs a "check for new events" button or similar.